LibraBFT/Impl/Consensus/RoundManager.agda

```diff
@@ -32,48 +33,76 @@ open RWST-do
 
 ------------------------------------------------------------------------------
 
+processCommitM : LedgerInfoWithSignatures → LBFT (List ExecutedBlock)
+processCommitM finalityProof = pure []
+
+-- IMPL-TODO: implement this
 processNewRoundEventM : Instant → NewRoundEvent → LBFT Unit
 processNewRoundEventM now nre = pure unit
 
 ------------------------------------------------------------------------------
 
-syncUpM               : Instant → SyncInfo → Author                   → LBFT (ErrLog ⊎ Unit)
-ensureRoundAndSyncUpM : Instant → Round    → SyncInfo → Author → Bool → LBFT (ErrLog ⊎ Bool)
+ensureRoundAndSyncUpM : Instant → Round    → SyncInfo → Author → Bool → LBFT (FakeErr ⊎ Bool)
 processProposalM      : Block                                         → LBFT Unit
-executeAndVoteM       : Block                                         → LBFT (ErrLog ⊎ VoteWithMeta)
+executeAndVoteM       : Block                                         → LBFT (FakeErr ⊎ Vote)
 
 -- external entry point
 -- TODO-2: The sync info that the peer requests if it discovers that its round
 -- state is behind the sender's should be sent as an additional argument, for now.
-processProposalMsgM : Instant → {- Author → -} ProposalMsg → LBFT Unit
-processProposalMsgM now {- from -} pm =
-  case pm ^∙ pmProposer of λ where
-    nothing        → pure unit -- log: info: proposal with no author
-    (just pAuthor) →
-      ensureRoundAndSyncUpM now (pm ^∙ pmProposal ∙ bRound) (pm ^∙ pmSyncInfo)
-                            pAuthor true >>= λ where
-      (inj₁ _)     → pure unit -- log: error: <propagate error>
-      (inj₂ true)  → processProposalM (pm ^∙ pmProposal)
-      (inj₂ false) → do
-        currentRound ← use (lRoundState ∙ rsCurrentRound)
-        pure unit -- log: info: dropping proposal for old round
+module processProposalMsgM (now : Instant) (pm : ProposalMsg) where
+  step₀ : LBFT Unit
+  step₁ : Author → LBFT Unit
+  step₂ : FakeErr ⊎ Bool → LBFT Unit
+
+  step₀ =
+    case pm ^∙ pmProposer of λ where
+      nothing → logInfo -- log: info: proposal with no author
+      (just pAuthor) → step₁ pAuthor
+  step₁ pAuthor =
+        ensureRoundAndSyncUpM now (pm ^∙ pmProposal ∙ bRound) (pm ^∙ pmSyncInfo)
+                              pAuthor true >>= step₂
+  step₂ r =
+        -- IMPL-DIFF: We use `ifM` to test whether the round of the proposal is
+        -- current, to take advantage of the obligations `RWST-weakestPre` generates.
+        case r of λ where
+        (inj₁ _) → logErr -- log: error: <propagate error>
+        (inj₂ pmCurrent) →
```

Comment:
The comment above addresses this (though since the new RWST monad is not a part of this PR, I had to convert the special-purpose case statements to ordinary case statements).

Every form of case analysis performed that isn't using one of the RWST operations will cause proofs to have to perform the same case analysis to proceed. I'll update the proofs in `chris-dijkstra-monad` your suggestion would break so we have something concrete to consider (perhaps I'm being overly aggressive in removing duplicate case analyzation in proofs).

Reply:
OK, we'll let this one simmer for a bit.